### PR TITLE
Move separation into skymap class and add unit test

### PR DESCRIPTION
**Pull request by OlgaVorokh**
Hi @cdeil! 
I've decided to start with a simpler task and move separation first. There was some test without asserts, so I just copied it and added assert in the middle -- not sure how to get ground truth for world=True. It looks good to me though so maybe this regression test will be enough.

World=false looks reasonable on a small 3x3 example.

### CODE REVIEW

File: gammapy/image/maps.py [hunk 1]
```diff
@@ -346,7 +348,7 @@ def reproject(self, refheader, mode='interp', *args, **kwargs):
             input_data=self.to_image_hdu(),
             output_projection=refheader,
             *args,
-            **kwargs,
+            **kwargs
```
Comment (OlgaVorokh) [inline gammapy/image/maps.py]: had py.test failure with comma

**Comment by adonath** [comment]
@OlgaVorokh Thanks for starting to work on this. I'll review this later.

**Comment by adonath** [comment]
I just did a [quick search for the `separation`](https://github.com/gammapy/gammapy/search?utf8=%E2%9C%93&q=separation%28&type=Code) function  in the gammapy code base and it turned out that this function isn't used at all (except in the test, which isn't even a test because it doesn't assert anything...)! Right now at every place separations or offsets are computed `SkyCoord.separation()` is used. So I think I'm in favour of removing the utility function `separation` completely, because it is just a left over, when astropy `SkyCoord` didn't exist yet. If there comes a use case in future, where we need e.g. separations in pixel coordinates, we can still add a `separation_pixel()` method to the `SkyMap` class.  

Instead I propose to slightly change the behaviour of `SkyMap.coordinates()`, so that `SkyCoords` are returned by default. Separation maps can then be computed like so:

```
center = SkyCoord(0, 0, frame='galactic' unit='deg')
center.separation(skymap.coordinates())
```
Which is still a nice and compact API. OK with you @cdeil and @joleroi?

**Comment by cdeil** [comment]
Well, there are cases like
https://github.com/gammapy/gammapy/blob/b6d3671a2ecdc36f6c12dcee6f5098c137d1c123/gammapy/cube/exposure.py#L39
where SkyCoord.separation should be used.

Another example should be where we're creating masks from circle sky regions.

I would propose to add a `skymap.separation(center)` utility method, even if it just calls `skymap.coordinates` followed by `SkyCoord.separation`.

In the future we'll probably extend it a bit with convenience flags like how to fill pixels beyond a max offset value (see the current exposure function).

@adonath - Is this acceptable to you, or do you strongly prefer not to add this convenience method?
(Final call up to you, just let @OlgaVorokh know exactly what to put in.)

@OlgaVorokh - FYI: https://github.com/gammapy/gammapy/blob/master/gammapy/utils/coordinates/celestial.py#L101 was added before `astropy.coordinates.SkyCoord` existed. It should be removed in this PR, we should go all-in on using Astropy stuff in Gammapy.

**Comment by adonath** [comment]
@cdeil I didn't understand the examples you mentioned, because `exposure_cube` already uses `SkyCoord.separation()` and I think there's no need to change it? 

Of course one could add things like max. separation or offset, but right now it is a single extra call to `np.clip()`, which is easily understandable and explicit code. Only when other cases appear and we duplicate this code at a few places, it's probably worth adding a convenience function.  Right now, I think, it isn't. So I'm still in favour of removing it.

**Comment by cdeil** [comment]
> @cdeil I didn't understand the examples you mentioned, because exposure_cube already uses SkyCoord.separation() and I think there's no need to change it?

Currently `exposure_cube` calls `np.meshgrid`. IMO it should call `cube.separation`, which in turn calls `cube.coordinates`, which in turn calls `np.meshgrid` (or equivalent). I.e. expose more things as methods on the image and cube class.

My point is that there will be many (~ 10 ?) places where we call `SkyMap.coordinates` followed by `SkyCoord.separation`. E.g. also for radially symmetric model evaluation.

I see `SkyMap` not just as a building block to implement algorithms, but also as a high-level key data structure that many users will use interactively or from scripts. So I think it's good to expose commonly used methods like separation and max offset handling in a convenience method.

@adonath - Can you summarise the changes to be done in this PR for @OlgaVorokh in a bullet or even better a GH task list?
Decide as you like, I won't argue further for a certain way to do it.

Just one more thing that hasn't been mentioned: I think the tests should be changed to use a large-FOV skymap with maybe 4 x 3 pixels covering a significant portion of the sky, so that spherical geometry is tested properly. Reference values could still be just obtained by copy & pasting the current output, adding it as a kind of regression test.

**Comment by adonath** [comment]
@cdeil Ok, I see you're point. But right now, I think, there's just no need for this. So I would introduce it, as soon as it's needed. Additionally I still think that both approaches (even when e.g. `np.clip()` is called after...) are easily understandable and lead to nice high level code for end-users as well. 

@OlgaVorokh My suggestion for this PR is as following:

- [ ] remove `separation` from `gammapy/image/utils.py`
- [ ] remove corresponding tests in `gammapy/image/tests/test_utils.py`
- [ ] remove the `coordinates` function from `gammapy/image/utils.py` (the only place it was still used was the `separation` function) 
- [ ] Optional: change the default behaviour of `SkyMap.coordinates()` to return a `SkyCoord`object. This might be a bit of work, because `.coordinates()` is used in many places in gammapy. So it might be better to do this in a separate PR. I leave it up to you, depending on the time you'd like to invest in this PR.

**Comment by OlgaVorokh** [comment]
I've removed the separation and it's tests.
The coordinates function in gammapy/image/utils.py is trickier, because it' used not only in separation, as @adonath mentioned, but also in gammapy.background.fov.fill_acceptance_image and gammapy.image.lookup_max/measure_containment/image_radius.

They all seem to call coordinates with the same default parameters, so I'll look into replacing it with astropy  coordinates or skymap.coordinates

**Comment by OlgaVorokh** [comment]
cc @cdeil 

I've looked into astropy coordinates [module](http://docs.astropy.org/en/stable/coordinates/), but I think I cannot replace `utils.coordinates` with it as this function is basically three astropy commands.  Should I attach it to `SkyMap` or leave as it is?

Now I'll work on optional part of changing `SkyMap.coordinates()` behavior.

**Comment by OlgaVorokh** [comment]
`Skycoords.coordinates` is used with three different mode parameters: `pix`, `skycoord` and `galaxy`
The `galaxy` is required only for `sherpa` convert methods(`to_sherpa_data2d`)
Each time the `pix` is used answer is immediately converted into `skymap`, so I guess we could remove `pix` mode from `skycoords.coordinates`, but we have to keep `galaxy` in order for `sherpa` unittests to work. 
Need advice on how to proceed.

**Comment by adonath** [comment]
@OlgaVorokh To completely remove the `coordinates` function from the `measure.py` sub-module, the functions have to be changed to take a `SkyMap` as input, instead of an `ImageHDU` object (and the same for the `fill_acceptance_image` function...). This requires a bit more of work, but  here are some hints:

`coordinates(image)` corresponds to `skymap.coordinates()` 
`image.data` corresponds to `skymap.data`
`image.header` corresponds to `skymap.wcs.to_header()`

For now I would keep the different modes (`galactic`, `pix`, etc.)., because I think it is still used at some other places (e.g.  `gammapy/image/profile.py`).

**Comment by OlgaVorokh** [comment]
@adonath I thought of constructing skymap inside the method, but I agree that it would make more sense to change method signatures instead.

Please bear with me while I am trying to understand coordinates :)

The old coordinates function had "world" coordinates, the skymap.coordinates has "pix", "skymap" and "galactic"
I can get equivalent to `coordinates(image)` result by doing
```
    lon, lat = skymap.coordinates(coord_type='pix')
    lon, lat = skymap.wcs.wcs_pix2world(x, y, 0)  # origin = 0 is convention for gammapy
    lon = np.where(lon > 180, lon - 360, lon)
```
So my question is -- is it something that skymap.coordinates() does or should I add "world" coord_type to `skymap.coordinates()`?
It looks very similar to 'galaxy' from the source code, but `pixel_to_skycoord()` is called from `SkyMap.coordinates` with default mode='all' instead of mode 'wcs'
```
    # Convert pixel coordinates to celestial coordinates
    if mode == 'all':
        lon, lat = wcs.all_pix2world(xp, yp, origin)
    elif mode == 'wcs':
        lon, lat = wcs.wcs_pix2world(xp, yp, origin)
    else:
        raise ValueError("mode should be either 'all' or 'wcs'")
```

Not sure how to correctly convert SkyCoord to lon/lat, 'galactic' modes accesses magic fields "l" and "b" and I've read that the attributes depend on a frame, and frame is set by wcs_to_celestial_frame(wcs) and god knows how it works and what fields do I need to get


To summarize, it would be easier to add new mode, but I am trying to avoid code duplication and understand how it works in the process

**Comment by cdeil** [comment]
Thanks! Merging this now.

I discussed with @adonath just now and the plan is that we improve `SkyMap.coordinates` to return a `SkyCoord`, and then it's possible to get the separation image with one line of code, so the need for a `SkyMap.separation` method is small, i.e. the plan at the moment is to **not** add one.
